sdk/tests/conformance/misc/views-with-offsets.html

```diff
@@ -0,0 +1,353 @@
+  }
+
+  ////
+
+  var tex2d = gl.createTexture();
+  gl.bindTexture(gl.TEXTURE_2D, tex2d);
+
+  gl.bindFramebuffer(gl.FRAMEBUFFER, fb);
+  gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, tex2d, 0);
+
+  testTexOrSubImage("texImage2D", fn2D);
+  testTexOrSubImage("texSubImage2D", fnSub2D);
+
+  ////
+
+  var tex3d = gl.createTexture();
+  gl.bindTexture(gl.TEXTURE_3D, tex3d);
+  gl.texImage3D(gl.TEXTURE_3D, 0, gl.RGBA, 1, 1, 1, 0, gl.RGBA, gl.UNSIGNED_BYTE, null);
+
+  gl.bindFramebuffer(gl.FRAMEBUFFER, fb);
+  gl.framebufferTextureLayer(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, tex3d, 0, 0);
+
+  testTexOrSubImage("texImage3D", fn3D);
+  testTexOrSubImage("texSubImage3D", fnSub3D);
+} while (false);
+
+
+do {
+  var compressedFormat = 0;
+  var compressedByteCount;
+
+  if (gl.getExtension("WEBGL_compressed_texture_s3tc")) {
+    var e = gl.getExtension("WEBGL_compressed_texture_s3tc");
+    compressedFormat = e.COMPRESSED_RGB_S3TC_DXT1_EXT;
+    compressedByteCount = 8;
+  } else if (gl.getExtension("WEBGL_compressed_texture_etc")) {
+    var e = gl.getExtension("WEBGL_compressed_texture_etc");
+    compressedFormat = e.COMPRESSED_RGB8_ETC2;
+    compressedByteCount = 8;
+  } else {
+    debug("No compressed texture format found. Skipping compressedTex(Sub)Image tests.");
+    break;
+  }
+
+  ////
+
+  var view = new Uint8Array(compressedByteCount+1);
+
+  var fn2D = function(viewOffset) {
+    gl.compressedTexImage2D(gl.TEXTURE_2D, 0, compressedFormat, 4, 4, 0,
+                            view, viewOffset, compressedByteCount);
+  };
+
+  var fnSub2D = function(viewOffset) {
+    gl.compressedTexImage2D(gl.TEXTURE_2D, 0, compressedFormat, 4, 4, 0,
+                            view, 0, compressedByteCount);
+    gl.compressedTexSubImage2D(gl.TEXTURE_2D, 0, 0, 0, 4, 4, compressedFormat,
+                               view, viewOffset, compressedByteCount);
+  };
+
+  var fn3D = function(viewOffset) {
+    gl.compressedTexImage3D(gl.TEXTURE_2D_ARRAY, 0, compressedFormat, 4, 4, 1, 0,
+                            view, viewOffset, compressedByteCount);
+  };
+
+  var fnSub3D = function(viewOffset) {
+    gl.compressedTexImage3D(gl.TEXTURE_2D_ARRAY, 0, compressedFormat, 4, 4, 1, 0,
+                            view, 0, compressedByteCount);
+    gl.compressedTexSubImage3D(gl.TEXTURE_2D_ARRAY, 0, 0, 0, 0, 4, 4, 1, compressedFormat,
+                               view, viewOffset, compressedByteCount);
+  };
+
+  ////
+
+  var testFunc = function(funcName, fnToTest) {
+    debug("");
+    debug(funcName);
+
+    if (!probeWithBadOffset(fnToTest, funcName))
+      return;
+
+    var viewLength = view.length;
+    var subViewLength = compressedByteCount;
+
+    for (var i = 0; i <= viewLength+1; i++) {
+      debug("offset=" + i);
+
+      fnToTest(i);
+      var effectiveViewLen = viewLength - i;
+
+      if (effectiveViewLen >= subViewLength) {
```

Comment:
This is subtly tricky. All of the existing compressed texture tests assume that the data length is exactly equal to the required amount, rather than being greater than or equal. Section 3.8.6 "Compressed Texture Images" of the ES 3.0.4 spec says:

"If the imageSize parameter is not consistent with the format, dimensions, and contents of the compressed image (too little or too much data), an INVALID_VALUE error results."

I'd be very concerned with changing this requirement at this stage of WebGL 2.0.0's release; doing so will require all of the compressed texture tests to be updated. It's also not clear whether drivers will handle the case where "too much" data is supplied, rather than exactly the right amount. I think this should be changed to:
```
  if (effectiveViewLen == subViewLength)
```